scipy/sparse/csgraph/_laplacian.py

```diff
@@ -83,72 +124,249 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+            md = _laplace_normed(graph, graph_sum, 1.0 / w)
+        if form == "function":
+            return md, w.astype(dtype, copy=False)
+        elif form == "lo":
+            m = _linearoperator(md, shape=graph.shape, dtype=dtype)
+            return m, w.astype(dtype, copy=False)
+    else:
+        if symmetrized:
+            md = _laplace_sym(graph, graph_sum)
+        else:
+            md = _laplace(graph, graph_sum)
+        if form == "function":
+            return md, diag.astype(dtype, copy=False)
+        elif form == "lo":
+            m = _linearoperator(md, shape=graph.shape, dtype=dtype)
+            return m, diag.astype(dtype, copy=False)
+
+
+def _laplacian_sparse(graph, normed, axis, copy, form, dtype, symmetrized):
+
+    if form != "array":
+        raise ValueError('{form} must be "array"'.format(form=repr(form)))
 
+    if dtype is None:
+        dtype = graph.dtype
 
-def _laplacian_sparse(graph, normed=False, axis=0,
-                      copy=True):
     needs_copy = False
-    if graph.format in ('lil', 'dok'):
+    if graph.format in ("lil", "dok"):
         m = graph.tocoo()
     else:
         m = graph
         if copy:
             needs_copy = True
+
+    if symmetrized:
+        m += m.T.conj()
+
     w = m.sum(axis=axis).getA1() - m.diagonal()
     if normed:
         m = m.tocoo(copy=needs_copy)
-        isolated_node_mask = (w == 0)
+        isolated_node_mask = w == 0
         w = np.where(isolated_node_mask, 1, np.sqrt(w))
         m.data /= w[m.row]
         m.data /= w[m.col]
         m.data *= -1
         m.setdiag(1 - isolated_node_mask)
     else:
-        if m.format == 'dia':
+        if m.format == "dia":
             m = m.copy()
         else:
             m = m.tocoo(copy=needs_copy)
         m.data *= -1
         m.setdiag(w)
-    return m, w
+
+    return m.astype(dtype, copy=False), w.astype(dtype)
 
 
-def _laplacian_dense(graph, normed=False, axis=0,
-                     copy=True):
+def _laplacian_dense_flo(graph, normed, axis, copy, form, dtype, symmetrized):
+
+    if dtype is None:
+        dtype = graph.dtype
+
     if copy:
         m = np.array(graph)
     else:
         m = np.asarray(graph)
 
+    if dtype is None:
+        dtype = m.dtype
+
+    graph_sum = m.sum(axis=axis)
+    graph_diagonal = m.diagonal()
+    diag = graph_sum - graph_diagonal
+    if symmetrized:
+        graph_sum += m.sum(axis=1 - axis)
+        diag = graph_sum - graph_diagonal - graph_diagonal
+
+    if normed:
+        isolated_node_mask = diag == 0
+        w = np.where(isolated_node_mask, 1, np.sqrt(diag))
+        if symmetrized:
+            md = _laplace_normed_sym(m, graph_sum, 1.0 / w)
+        else:
+            md = _laplace_normed(m, graph_sum, 1.0 / w)
+        if form == "function":
+            return md, w.astype(dtype, copy=False)
+        elif form == "lo":
+            m = _linearoperator(md, shape=graph.shape, dtype=dtype)
+            return m, w.astype(dtype, copy=False)
```

Comment:
Looks like we're missing the final `else:` branch here.

```
else:
  raise ValueError(f"Invalid form: {repr(form)}")
```

Reply:
done